### ISSUE TITLE
Provide a way to clone Context

### ISSUE THREAD

**Issue by harnash** [go-playground/lars#20]
Hi.
I'm in a process of adding OpenTracing to our app which uses lars. To do that I've created custom context to simplify some operations. The way the OpenTracing API works is that it injects some information about the spans (https://github.com/opentracing/specification/blob/master/specification.md) into the Context object to be propagated down into the call chain (or via headers to other services being called in the process).

The problem arises when I create multiple spans in the same request. My function looks more or less like this:
```go
func (c CustomContext) StartSpan(operationName string) (opentracing.Span, *CustomContext) {
	span, newCtx := opentracing.StartSpanFromContext(c.Context(), operationName)

	c.WithContext(newCtx)
	return span, &c
}
```
The obvious problem is that when it is being called more than once the current Context is being altered by every call of this function. What I want is to return copy of the context and use it in subsequent functions without altering the "parent" Context.
```go
func handler(c *CustomContext) {
   spanA, newCtx := c.StartSpan("spanA")
   defer spanA.Finish()
   subFunc(newCtx)

   spanC, _ := c.StartSpan("spanC")
   defer spanC.FInish()
}

func subFunc(c *CustomContext) {
   spanB, _ := c.StartSpan("spanB")
   // ...
   defer spanB.FInish()

}
```
this currently result in spans like this:
```
               ------------
               | RootSpan |
               ------------
                  | 
            ---------
            | SpanA |
            ---------
                   |
                ---------
                | SpanB |
                ---------
                     |
                ---------
                | SpanC |
                ---------
```
proper hierarchy should look like this:
```
               ------------
               | RootSpan |
               ------------
                 |       |
            ---------  ---------
            | SpanA |  | SpanC |
            ---------  ---------
                   |
                ---------
                | SpanB |
                ---------
```
The only solution to this is I can see right now is to:
a) create a full deep clone of the lars.Context struct
b) ditch lars.Context in the handler and operate on standard Context (accessible via lars.Context.Context()) - but that would mean that in functions that needs to operate on lars.Context I would need to pass two of them (lars.Context and std.Context) which is not ideal

I'm open to any suggestions how to deal with it.

**Comment by harnash** [comment]
There may be a third option:
c) https://github.com/containerum/gintracing/blob/master/gintracing.go but I would rather use functions provided by opentracing-go lib than hack it around

Note: it still won't fix the context propagation...

**Comment by deankarn** [comment]
Hey @harnash the way I understand it is that `context.Context` is meant to be used, for lack of a better term, in a straight line so when it deviates to B like in your example, that throws some wrinkles into it.

They are not insurmountable though, the only way ATM I can think of achieving what you desire is that in this sample:
```go
func handler(c *CustomContext) {
   spanA, newCtx := c.StartSpan("spanA")
   defer spanA.Finish()
   subFunc(newCtx)

   spanC, _ := c.StartSpan("spanC")
   defer spanC.FInish()
}
``` 
The context used with a needs to be assigned to a variable to then be used with C knowing that sub-function B will change it.

so essentially 2 methods on the `*CustomContext`
```go
func (c *CustomContext) StartSpan(operationName string) opentracing.Span {
	span, newCtx := opentracing.StartSpanFromContext(c.Context(), operationName)
	c.WithContext(newCtx)
	return span
}

func (c *CustomContext) StartSpanWithContext(operationName string,ctx context.Context) opentracing.Span {
	span, newCtx := opentracing.StartSpanFromContext(ctx, operationName)
	c.WithContext(newCtx)
	return span
}
```

then your code can become
```go
func handler(c *CustomContext) {
   spanA := c.StartSpan("spanA")
   defer spanA.Finish()
   ctx := c.Context()
   subFunc(c)

   spanC, _ := c.StartSpanWithContext("spanC", ctx)
   defer spanC.FInish()
}

func subFunc(c *CustomContext) {
   spanB, _ := c.StartSpan("spanB")
   // ...
   defer spanB.FInish()

}
```

There may be other ways, but this seems the simplest; I haven't looked too deeply into open tracing so if this is way off I apologize.

Please let me know if this helps at all :)

**Comment by harnash** [comment]
Thanks for a quick reply @joeybloggs.

This is exactly my solution b) this will probably work but there are some caveats to this:

1. We are operating on two Context's (one kept in `CustomContext` and second `context.Context` passed by from any handler) causing some confusion when to use one and when the other. In some cases we would need to make sure that information accessible in one is also accessible in other if we limit the number of contexts being passed down to functions.
2. In functions that need both (like some error handling etc.) when we need to generate responses or look into `http.Request` object there will be need of passing two contexts and make sure we use the right one when dealing with tracing code.

What actually I need is new version of `context.Context` being passed down the call chain which will change whenever I create child spans. I might end up keeping copy of `context.Context` in my `CustomContext` just for tracing purposes and try to hide it through some utility methods but it is still a hack 👿 
I know that "cloning" (at least the `context.Context` part) can have memory/performance hit so I'm not pushing - just trying to find sensible solution as I see others are struggling to do it right too.

**Comment by harnash** [comment]
What would be nice it have something like `http.Request`'s `WithContext` which will return new `lars.Ctx` with modified request context. It would need to replace `lars.Ctx` underlying `http.Request` with a new one and newly applied context. What do you think?

**Comment by harnash** [comment]
What if `Ctx.WithContext()` would return modified `lars.Ctx` instead of modifying it in-place?

**Comment by harnash** [comment]
I've found another workaround...
Previously I had my `CustomContext` defined like this:
```go
type CustomContext struct {
	*lars.Ctx
	logger *logrus.Entry
	Tracer opentracing.Tracer
}
```
when I've changed it to this:
```go
type CustomContext struct {
	lars.Ctx
	logger *logrus.Entry
	Tracer opentracing.Tracer
}
```
it appears to be working as expected... I'm not 100% sure of the side effects it may generate but since I'm passing around `*CustomContext` it should be OK right?

**Comment by deankarn** [comment]
@harnash took a quick look and it looks like it should work fine, I would run test though to ensure a value you set in the context.Context say in a middleware is accessible form subsequent middleware and the final handler, just to be safe

**Comment by harnash** [comment]
@joeybloggs Sure. We will test it. To make sure I'll do some benchmarks to see if memory footprint is acceptable and performance is not degraded too much.

**Comment by harnash** [comment]
I'm closing this ticket since I've found a workable solution

**Comment by deankarn** [comment]
Just out of curiosity how did the benchmarking and testing go?

**Comment by harnash** [comment]
Testing went well so far. I did some local testing on our server and with tracing I got increase from ~17ms to ~24ms on a quite simple resource (static resource). This is measured with tracing span logging on and 100% of spans being reported. It goes down to ~20ms with 1% sampling and logging turned off. I didn't do synthetic benchmarks since results were satisfactory for us for now. I might do that later but for now I need to focus on other stuff.
Memory went up but that was expected since we changed the way we use context (we propagate some additional data now).
